src/extensions/looseassembly/Client/StrongNameInfo.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Extensions.LooseAssembly.Client
+{
+    /// <summary>
+    /// Encapsulates a strong name "family" identity (name and public key token).
+    /// </summary>
+    public class StrongNameInfo : IEquatable<StrongNameInfo>
+    {
+        public string Name { get; }
+
+        public PublicKeyToken? PublicKeyToken { get; }
+
+        public StrongNameInfo(string name, PublicKeyToken publicKeyToken)
+        {
+            Name = name;
+            PublicKeyToken = publicKeyToken;
+        }
+
+        public StrongNameInfo(string name)
+        {
+            Name = name;
+            PublicKeyToken = null;
+        }
+
+        public bool Equals(StrongNameInfo? other) => Name.Equals(other?.Name, StringComparison.OrdinalIgnoreCase) && PublicKeyToken == other?.PublicKeyToken;
+
+        public override int GetHashCode()
+        {
+            return HashCode.Combine(Name, PublicKeyToken);
+        }
+
+        public override bool Equals(object? obj) => obj switch
+        {
+            StrongNameInfo sn => Equals(sn),
+            _ => false,
+        };
+
+        public static StrongNameInfo? GetStrongName(Stream stream)
+        {
+            try
+            {
+                using var peReader = new PEReader(stream);
+
+                if (!peReader.HasMetadata)
+                {
+                    return default;
+                }
+
+                var reader = peReader.GetMetadataReader();
+
+                if (!reader.IsAssembly)
+                {
+                    return default;
+                }
+
+                var assemblyName = reader.GetAssemblyDefinition().GetAssemblyName();
```

Comment:
It's worth noting that GetAssemblyDefinition can fail (return null?) for modules. I haven't looked at the behavior here. It's rare, but it's probably worth hardening for.

Reply:
Would the call to `IsAssembly` be sufficient ( a few lines up)?

Reply:
Per the code (https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Metadata/MetadataReader.cs,0fb3765222daa266) it looks like that's the check that would throw.

I've already got that. Is there any other hardening you're thinking of here?

Reply:
Yeah, you got it. I didn't notice the IsAssembly check. That's sufficient here. And IsAssembly is doing the equivalent number of rows check that I was previously doing. Seems fine.